File: 백준/Gold/9663. N－Queen/N－Queen.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    public static int N;
    public static int ans = 0;
    public static int[] v1,v2,v3;

    public static void dfs(int n) {
        if (n == N) {
            ans ++;
            return;
        }

        for (int j = 0; j < N; j++) {
            if (v1[j] == 0 && v2[n - j + (N - 1)] == 0 && v3[n + j] == 0) {
                v1[j] = v2[n - j + (N - 1)] = v3[n + j] = 1;
                dfs(n + 1);
                v1[j] = v2[n - j + (N - 1)] = v3[n + j] = 0;
            }
        }
    }

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        ans = 0;

        v1 = new int[N];
        v2 = new int[2 * N - 1];
        v3 = new int[2 * N - 1];

        dfs(0);
        System.out.println(ans);
    }
}

File: 백준/Gold/9663. N－Queen/N－Queen.py
import sys

input = sys.stdin.readline


def dfs(n):
    global ans
    if n == N:
        ans += 1
        return

    for i in range(N):
        if v1[i] == 0 and v2[n - i] == 0 and v3[n + i] == 0:
            v1[i] = v2[n - i] = v3[n + i] = 1
            dfs(n + 1)
            v1[i] = v2[n - i] = v3[n + i] = 0


N = int(input())
ans = 0
v1 = [0] * N
v2 = [0] * N * 2
v3 = [0] * N * 2
dfs(0)
print(ans)